File: iOS/AppFactory/AppFactory/Http/Requester/EngineerDetailRequester.swift
//
//  EngineerDetailRequester.swift
//  AppFactory
//
//  Created by Leapfrog-Software on 2018/04/09.
//  Copyright © 2018年 Leapfrog-Inc. All rights reserved.
//

import Foundation

struct EngineerDetailWorkData {
    
    let id: String
    let name: String
    let os: String
    let language: String
    let function: String
    let cost: Int
    let evaluate: Int
    let source: Bool
    
    init?(data: Dictionary<String, Any>) {

        guard let id = data["id"] as? String else {
            return nil
        }
        self.id = id
        self.name = data["name"] as? String ?? ""
        self.os = data["os"] as? String ?? ""
        self.language = data["language"] as? String ?? ""
        self.function = data["function"] as? String ?? ""
        self.cost = data["cost"] as? Int ?? 0
        self.evaluate = data["evaluate"] as? Int ?? 0
        self.source = data["source"] as? Bool ?? false
    }
}

struct EngineerDetailResponseData {
    
    let id: String
    let name: String
    let area: String
    let organization: OrganizationType
    let evaluate: Int
    let profile: String
    let works: [EngineerDetailWorkData]
    
    init?(data: Dictionary<String, Any>) {

        guard let id = data["id"] as? String else {
            return nil
        }
        self.id = id
        
        self.name = data["name"] as? String ?? ""
        self.area = data["area"] as? String ?? ""
        self.organization = OrganizationType.create(with: data["organization"] as? Int ?? 0)
        self.evaluate = data["evaluate"] as? Int ?? 0
        self.profile = data["profile"] as? String ?? ""
        self.works = (data["works"] as? Array<Dictionary<String, Any>>)?.flatMap { EngineerDetailWorkData(data: $0) } ?? []
    }
}

class EngineerDetailRequester {
    
    class func get(id: String, completion: @escaping ((EngineerDetailResponseData?) -> ())) {
        
        var url = Constants.ServerRootUrl + Constants.ServerApiFile
        url += "?"
        url += ("command=getengineerdetail")
        url += "&"
        url += ("id=" + id)
        
        ApiManager.get(url: url, completion: { result, data in
            if result, let dic = data {
                completion(EngineerDetailResponseData(data: dic))
            } else {
                completion(nil)
            }
        })
    }
}

File: Android/app/src/main/java/leapfrog_inc/appfactory/Http/Requester/EngineerDetailRequester.java
package leapfrog_inc.appfactory.Http.Requester;

import android.util.Base64;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;

import leapfrog_inc.appfactory.Function.Constants;
import leapfrog_inc.appfactory.Http.Enum.CostType;
import leapfrog_inc.appfactory.Http.Enum.OrganizationType;
import leapfrog_inc.appfactory.Http.Enum.WorksType;
import leapfrog_inc.appfactory.Http.HttpManager;

/**
 * Created by Leapfrog-Software on 2018/04/16.
 */

public class EngineerDetailRequester {

    public static class EngineerDetailWorkData {

        public String id;
        public String name;
        public String os;
        public String language;
        public String function;
        public int cost;
        public int evaluate;
        public boolean source;

        static public EngineerDetailWorkData create(JSONObject json) {

            try {
                String id = json.getString("id");
                String name = json.getString("name");
                String os = json.getString("os");
                String language = json.getString("language");
                String function = json.getString("function");
                int cost = json.getInt("cost");
                int evaluate = json.getInt("evaluate");
                boolean source = json.getBoolean("source");

                EngineerDetailWorkData data = new EngineerDetailWorkData();
                data.id = id;
                data.name = name;
                data.os = os;
                data.language = language;
                data.function = function;
                data.cost = cost;
                data.evaluate = evaluate;
                data.source = source;

                return data;

            } catch(Exception e) {}

            return null;
        }
    }

    public static class EngineerDetailResponseData {

        public String id;
        public String name;
        public String area;
        public OrganizationType organization;
        public int evaluate;
        public String profile;
        public ArrayList<EngineerDetailWorkData> works = new ArrayList<EngineerDetailWorkData>();

        static public EngineerDetailResponseData create(JSONObject json) {

            try {
                EngineerDetailResponseData data = new EngineerDetailResponseData();
                data.id = json.getString("id");
                data.name = json.getString("name");
                data.area = json.getString("area");
                data.organization = OrganizationType.create(json.getInt("organization"));
                data.evaluate = json.getInt("evaluate");
                data.profile = json.getString("profile");

                JSONArray works = json.getJSONArray("works");
                for (int i = 0; i < works.length(); i++) {
                    EngineerDetailWorkData workData = EngineerDetailWorkData.create(works.getJSONObject(i));
                    if (workData != null) {
                        data.works.add(workData);
                    }
                }
                return data;

            } catch (Exception e) {}

            return null;
        }
    }

    public static void get(String id, final EngineerDetailRequesterCallback callback){

        HttpManager httpManager = new HttpManager(new HttpManager.HttpCallback() {
            @Override
            public void didReceiveData(boolean result, String data) {

                if (result) {
                    try {
                        EngineerDetailResponseData responseData = EngineerDetailResponseData.create(new JSONObject(data));
                        if (responseData != null) {
                            callback.didReceiveData(true, responseData);
                            return;
                        }
                    } catch (Exception e) {}
                }
                callback.didReceiveData(false, null);
            }
        });

        StringBuffer url = new StringBuffer();
        url.append(Constants.ServerApiUrl);
        url.append("?");
        url.append("command=getengineerdetail");
        url.append("&");
        url.append(("id=" + id));

        httpManager.execute(url.toString(), "GET", null);
    }

    public interface EngineerDetailRequesterCallback {
        void didReceiveData(boolean result, EngineerDetailResponseData response);
    }
}

